core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2018, STMicroelectronics
+ */
+
+#include <arm.h>
+#include <sm/optee_smc.h>
+#include <sm/sm.h>
+#include <string.h>
+#include <stm32_util.h>
+
+#include "stm32mp1_smc.h"
+
+#define SM_NSEC_CTX_OFFSET(_reg)	(offsetof(struct sm_nsec_ctx, _reg) - \
+					 offsetof(struct sm_nsec_ctx, r0))
+#define SMC_ARGS_OFFSET(_reg)		offsetof(struct smc_args, _reg)
+
+static enum sm_handler_ret sip_service(struct sm_ctx *ctx __unused,
+				       struct smc_args *args)
+{
+	switch (OPTEE_SMC_FUNC_NUM(args->a0)) {
+	case STM32_SIP_SVC_FUNC_CALL_COUNT:
+		args->a0 = STM32_SIP_SVC_FUNCTION_COUNT;
+		break;
+	case STM32_SIP_SVC_FUNC_VERSION:
+		args->a0 = STM32_SIP_SVC_VERSION_MAJOR;
+		args->a1 = STM32_SIP_SVC_VERSION_MINOR;
+		break;
+	case STM32_SIP_SVC_FUNC_UID:
+		args->a0 = STM32_SIP_SVC_UID_0;
+		args->a1 = STM32_SIP_SVC_UID_1;
+		args->a2 = STM32_SIP_SVC_UID_2;
+		args->a3 = STM32_SIP_SVC_UID_3;
+		break;
+	default:
+		return SM_HANDLER_PENDING_SMC;
+	}
+
+	return SM_HANDLER_SMC_HANDLED;
+}
+
+enum sm_handler_ret sm_platform_handler(struct sm_ctx *ctx)
+{
+	struct smc_args *args = (void *)&ctx->nsec.r0;
+
+	COMPILE_TIME_ASSERT(SM_NSEC_CTX_OFFSET(r0) == SMC_ARGS_OFFSET(a0) &&
+			    SM_NSEC_CTX_OFFSET(r1) == SMC_ARGS_OFFSET(a1) &&
+			    SM_NSEC_CTX_OFFSET(r2) == SMC_ARGS_OFFSET(a2) &&
+			    SM_NSEC_CTX_OFFSET(r3) == SMC_ARGS_OFFSET(a3) &&
+			    SM_NSEC_CTX_OFFSET(r4) == SMC_ARGS_OFFSET(a4) &&
+			    SM_NSEC_CTX_OFFSET(r5) == SMC_ARGS_OFFSET(a5) &&
+			    SM_NSEC_CTX_OFFSET(r6) == SMC_ARGS_OFFSET(a6) &&
+			    SM_NSEC_CTX_OFFSET(r7) == SMC_ARGS_OFFSET(a7));
```

Comment:
Related to comment https://github.com/OP-TEE/optee_os/pull/3736/commits/58b3a57dfcbcb397387de872b9e7f1db6a2a9ac6#r406267428 in the P-R: this assertion was to ensure `struct sm_nsec_ctx`:`r0`..`r7`  do match fields struct thread_smc_args`:`a0`..`a7`. 
Is it really necessary?

There is a cast in [**sm.c**](https://github.com/OP-TEE/optee_os/blob/3.8.0/core/arch/arm/sm/sm.c#L18-L36) that implicitly expect this mapping:
```c
uint32_t sm_from_nsec(struct sm_ctx *ctx)
{
	uint32_t *nsec_r0 = (uint32_t *)(&ctx->nsec.r0);

	(...)

		smc_std_handler((struct thread_smc_args *)nsec_r0, &ctx->nsec);
}
```
Do you suggest to do the same here, removing the assertion.

Reply:
Yes, I think that's safe (to remove the assertion).